CS4850Group5A/VirtuDoc
# Database Glue

# Summary
- Added JDBC driver dependencies for H2 and MySQL.
- Added MySQL to local development stack.
- Added MySQL service to GHA environment.
- Added sample entity to test database connection and to demonstrate usage of the JPA Entity paradigm.
- Changed `default` application profile to use H2 database.
- Changed Docker Compose application profile to `dev-managed` instead of `default`.

Closes #2 

# How to Test
- Launch the Docker Compose stack. Verify that you can visit [http://localhost:8080/sampleentity/all](http://localhost:8080/sampleentity/all) and that you see the text `[]` in your browser.
- If you have a SQL client installed, connect to the MySQL server in the stack at `mysql://virtudoc:33306`. The username is `root` and the password is `virtudoc`. The database is `virtudoc`. Add an entry to the SampleEntity table and visit the above URL again. Some JSON should be returned with your SampleEntity object.
- Run the JUnit test suite and confirm that all tests pass locally using the H2 database.

# Additional Comments
## H2 Database
If you run the application natively (e.g. JUnit tests or running without the Docker Compose stack), the app is now configured to use the H2 database. This is an in-memory SQL database, meaning that when the application closes, the entire database is deleted. This helps ensure that unit tests have a clean environment to perform their tests independent of your development dataset. When running in Docker Compose, GitHub Actions, or on Heroku, MySQL is used.

## JPA + Spring Data ORM
This PR introduces the JPA Entity architecture. All objects in the database are declared in Java code using the Entity class decorator. This means that all objects can be defined as POJO(plain-old Java objects), without the need to write SQL code. This also makes the application work across any JDBC database with no code changes. Spring Data also handles connection pooling, so we won't accidentally go over our connection limit of 10. Finally, this ORM also protects against SQL-injection attacks and other common security issues that can arise when dealing with raw database connections.

## Known Issues
- If you are using IntelliJ to manage your Docker Compose stack, you will need to open the configuration and add in the new `db` service.
- If you receive "Whitelabel error page", you may have old artifacts in your target directory. Shut down the stack, delete everything in your `target/` folder, then rebuild the stack and try again.
